subtests/docker_cli/deferred_deletion/deferred_deletion.py

```diff
@@ -0,0 +1,153 @@
+r"""
+Summary
+---------
+
+Test deferred deletion: device mapper will not delete a docker
+filesystem if it's still in use.
+
+This test is only applicable on RHEL 7.4+ systems in which:
+
+  1) kernel is 3.10.0-632 or above; and
+  2) The sysctl knob /proc/sys/fs/may_detach_mounts exists and is 1; and
+  3) dockerd unit file DOES NOT include MountFlags=slave; and
+  4) docker daemon is configured to use devicemapper storage; and
+  5) docker daemon is run with --storage-opts dm.use_deferred_deletion=true
+
+As a precondition for running this test, we check only conditions 4 & 5,
+aborting with TestNAError if either is false. We trust docker-storage-setup
+to only enable #5 if all preconditions are met. Anyone running a system
+with 4 & 5 true but 1, 2, or 3 false deserves what they get.
+
+Operational Summary
+----------------------
+
+#. Run a docker container
+#. Find out the local (host) mountpoint of its root filesystem. cd into it.
+#. Read the Deferred-Deletion count from 'docker info'.
+#. Remove the container.
+#. Read the Deferred-Deletion count; confirm that it has grown by one.
+#. cd back out of the container rootfs
+#. Read the Deferred-Deletion count; confirm that it has gone back down.
+
+"""
+
+import os
+from autotest.client import utils
+import dockertest.docker_daemon
+from dockertest import subtest
+from dockertest.containers import DockerContainers
+from dockertest.dockercmd import AsyncDockerCmd, DockerCmd
+from dockertest.images import DockerImage
+from dockertest.output import DockerInfo, mustpass
+from dockertest.xceptions import DockerTestNAError
+
+
+class deferred_deletion(subtest.Subtest):
+
+    def initialize(self):
+        super(deferred_deletion, self).initialize()
+        dockerinfo = DockerInfo()
+        storage_driver = dockerinfo['Storage Driver']
+        if storage_driver != 'devicemapper':
+            raise DockerTestNAError("This test is only applicable when using"
+                                    " the devicemapper storage driver; host"
+                                    " system is using %s" % storage_driver)
+
+        docker_cmdline = dockertest.docker_daemon.cmdline()
+        if "dm.use_deferred_deletion=true" not in str(docker_cmdline):
+            raise DockerTestNAError("deferred deletion not enabled in"
+                                    " docker daemon command-line options.")
+
+    def run_once(self):
+        super(deferred_deletion, self).run_once()
+
+        self._start_idle_container()
+
+        # cd into its directory. This prevents devicemapper from reaping it.
+        os.chdir(self._container_mountpoint)
```

Comment:
Impatient, short-tempered future maintainers may become frustrated that  ``_container_mountpoint`` doesn't appear when they scroll up.

Since it's only referenced once, does it really make sense to be a property?  Would it be simpler to ``self.start_idle_container()`` from ``initialize()`` and then ``self.stuff['container_mountpoint'] = self.idler_mountpoint()``?

I dunno, it's okay either way, okay, forget I even brought it up, damn, too late.